File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/javascript/exercises/practice/forth/forth.js
//
// This is only a SKELETON file for the 'Forth' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Forth {
  constructor() {
    this._stack = [];
    this._customWords = new Map();
  }

  evaluate(input) {
    const tokens = input.toUpperCase().split(/[\s]+/).filter(t => t !== '');
    let definingCustom = false;
    let customWordName = '';
    let customWordDef = [];

    let i = 0;
    while (i < tokens.length) {
      const token = tokens[i];
      i++;
      
      if (definingCustom) {
        if (token === ';') {
          this._customWords.set(customWordName, customWordDef);
          definingCustom = false;
        } else if (!customWordName) {
          if (/^\d+$/.test(token)) {
            throw new Error('Invalid word name');
          }
          customWordName = token;
        } else {
          customWordDef.push(token);
        }
        continue;
      }

      if (token === ':') {
        definingCustom = true;
        customWordName = '';
        customWordDef = [];
        continue;
      }

      // Handle numbers
      if (/^\d+$/.test(token)) {
        this._stack.push(parseInt(token, 10));
        continue;
      }

      // Resolve custom words first
      if (this._customWords.has(token)) {
        const def = this._customWords.get(token);
        tokens.splice(i, 1, ...def);  // Replace current token with definition
        i = 0;  // Restart processing from beginning of modified tokens
        continue;
      }

      // Built-in operations
      switch(token) {
        case '+': this._checkStackSize(2); this._stack.push(this._stack.pop() + this._stack.pop()); break;
        case '-': this._checkStackSize(2); { const a = this._stack.pop(); this._stack.push(this._stack.pop() - a); } break;
        case '*': this._checkStackSize(2); this._stack.push(this._stack.pop() * this._stack.pop()); break;
        case '/': this._checkStackSize(2); { const a = this._stack.pop(); if (a === 0) throw new Error('Division by zero'); this._stack.push(Math.floor(this._stack.pop() / a)); } break;
        case 'DUP': this._checkStackSize(1); this._stack.push(this._stack[this._stack.length - 1]); break;
        case 'DROP': this._checkStackSize(1); this._stack.pop(); break;
        case 'SWAP': this._checkStackSize(2); { const a = this._stack.pop(), b = this._stack.pop(); this._stack.push(a, b); } break;
        case 'OVER': this._checkStackSize(2); this._stack.push(this._stack[this._stack.length - 2]); break;
        default: throw new Error(`Unknown word: ${token}`);
      }
    }
  }

  _checkStackSize(n) {
    if (this._stack.length < n) {
      throw new Error('Stack underflow');
    }
  }

  get stack() {
    return [...this._stack];
  }
}

File: src/2025-03-04-12-20-21--deepseek-vr-0304/go/exercises/practice/forth/forth.go
package forth

import (
	"fmt"
	"strconv"
	"strings"
)

func Forth(input []string) ([]int, error) {
	stack := []int{}
	customWords := make(map[string][]string)

	for _, line := range input {
		tokens := strings.Fields(line)
		if len(tokens) == 0 {
			continue
		}

		if strings.ToLower(tokens[0]) == ":" {
			err := processDefinition(tokens, customWords)
			if err != nil {
				return nil, err
			}
			continue
		}

		err := evaluateTokens(tokens, customWords, &stack)
		if err != nil {
			return nil, err
		}
	}

	return stack, nil
}

func processDefinition(tokens []string, customWords map[string][]string) error {
	if len(tokens) < 4 || strings.ToLower(tokens[len(tokens)-1]) != ";" {
		return fmt.Errorf("invalid definition")
	}

	wordName := strings.ToLower(tokens[1])
	if _, err := strconv.Atoi(wordName); err == nil {
		return fmt.Errorf("invalid word name")
	}

	definition := tokens[2 : len(tokens)-1]
	customWords[wordName] = definition
	return nil
}

func evaluateTokens(tokens []string, customWords map[string][]string, stack *[]int) error {
	for _, token := range tokens {
		tokenLower := strings.ToLower(token)
		if num, err := strconv.Atoi(token); err == nil {
			*stack = append(*stack, num)
			continue
		}

		if def, ok := customWords[tokenLower]; ok {
			err := evaluateTokens(def, customWords, stack)
			if err != nil {
				return err
			}
			continue
		}

		switch tokenLower {
		case "+", "-", "*", "/":
			err := arithmeticOp(tokenLower, stack)
			if err != nil {
				return err
			}
		case "dup":
			if len(*stack) < 1 {
				return fmt.Errorf("stack underflow")
			}
			*stack = append(*stack, (*stack)[len(*stack)-1])
		case "drop":
			if len(*stack) < 1 {
				return fmt.Errorf("stack underflow")
			}
			*stack = (*stack)[:len(*stack)-1]
		case "swap":
			if len(*stack) < 2 {
				return fmt.Errorf("stack underflow")
			}
			(*stack)[len(*stack)-2], (*stack)[len(*stack)-1] = (*stack)[len(*stack)-1], (*stack)[len(*stack)-2]
		case "over":
			if len(*stack) < 2 {
				return fmt.Errorf("stack underflow")
			}
			*stack = append(*stack, (*stack)[len(*stack)-2])
		default:
			return fmt.Errorf("undefined word %s", token)
		}
	}
	return nil
}

func arithmeticOp(op string, stack *[]int) error {
	if len(*stack) < 2 {
		return fmt.Errorf("stack underflow")
	}
	b := (*stack)[len(*stack)-1]
	a := (*stack)[len(*stack)-2]
	*stack = (*stack)[:len(*stack)-2]

	var result int
	switch op {
	case "+":
		result = a + b
	case "-":
		result = a - b
	case "*":
		result = a * b
	case "/":
		if b == 0 {
			return fmt.Errorf("division by zero")
		}
		result = a / b
	default:
		return fmt.Errorf("invalid operator")
	}
	*stack = append(*stack, result)
	return nil
}
